### PR TITLE
dai: assign link DMA channel at run-time

### PR DESCRIPTION
**Pull request by ranj063**
The recommended HDA HW programming sequence for setting the DMA format requires that the link DMA and host DMA channels be coupled before setting the format. This change means that host DMA or link DMA channels be reserved even if only one is used.

Statically assigned link DMA channels would mean that all the corresponding host DMA channels will need to be reserved, leaving only a few channels available at run-time. So, the suggestion here is to switch to dynamically assigning both host DMA channels and link DMA channels are run-time.

The host DMA channel is assigned when the pcm is opened as before. While choosing the link DMA channel, if the host DMA channel corresponding to the link DMA channel is already taken, the proposed method checks to make sure that the BE is connected to the FE that has been assigned this host DMA channel. For hostless streams, the host DMA channel is marked as opened to ensure it isn't assigned to another stream. After assigning the link DMA channel, a DAI_CONFIG IPC is sent to the FW to set the chosen link DMA channel.

This change should support N:M FE <->BE configurations as well.

The corresponding kernel change for this PR is https://github.com/thesofproject/linux/pull/895

### CODE REVIEW

**Comment by xiulipan** [comment]
@ranj063 
manual trigger a test with https://github.com/thesofproject/linux/pull/895
test logs on https://sof-ci.01.org/sofpr/PR1354/build1833/

**Comment by ranj063** [comment]
Thanks @xiulipan

File: src/include/uapi/abi.h [hunk 1]
```diff
@@ -51,8 +51,8 @@
 #define __INCLUDE_UAPI_ABI_H__
 
 /** \brief SOF ABI version major, minor and patch numbers */
-#define SOF_ABI_MAJOR 3
-#define SOF_ABI_MINOR 6
+#define SOF_ABI_MAJOR 4
```
Comment (lgirdwood) [inline src/include/uapi/abi.h]: I can't see where the ABI is changed ? Is there a patch missing from the PR ?

Comment (ranj063) [reply src/include/uapi/abi.h]: @lgirdwood there are no changes to the IPC structures in this PR but I changed the ABI because the way the link DMA channels are set is different. So kernel that sets the link DMA channel dynamically would expect the FW to do the same and if the FW is older where the link DMA channels are set statically, all HDA usecases would fail. So they are incompatible. I didnt know how else to call out this mismatch.

Comment (lgirdwood) [reply src/include/uapi/abi.h]: ok, but I also dont see any changes to the IPC programming flow (including in the kernel PR), this looks more like a dependency on register programming flow depending on how FW allocates HDA DMA channels ? If so, it's probably better to advertise this (using boot extended metadata) so the driver can then decide which programming version to use. This also means that 1:N could still be supported in FW/driver (and re-used when HW supports it).

Comment (ranj063) [reply src/include/uapi/abi.h]: @lgirdwood I could advertise this with boot extended metadata but I don't understand how this would help the 1:N case. This scenario would be broken even with the static allocation scheme that we have today.

Comment (lgirdwood) [reply src/include/uapi/abi.h]: @ranj063 ok, so you mean that "change will not work with 1:N FE <-> BE connections" is even true today without this change.

Comment (ranj063) [reply src/include/uapi/abi.h]: @lgirdwood yes thats correct

Comment (ranj063) [reply src/include/uapi/abi.h]: @lgirdwood the boot ext metadata would handle the backward compatibility between a new kernel and an older FW. But do you have any suggestions for newer FW and older kernel? 
We send the link_dma_ch as part of the DAI config IPC. In the case of dynamic allocation, this would still be there but just that we ignore the link_dma_ch field which will always be 0 from the driver.

Comment (ranj063) [reply src/include/uapi/abi.h]: @lgirdwood answering my question myself. I think I've found a way to do this now. I will update the PR soon

**Comment by ranj063** [comment]
@lgirdwood @plbossart I've updated both the kernel and FW PR's with minor ABI updates with the addition of reserved member in the hda params IPC structure. Let me know if it is too much and I'll trim it down. 

@lgirdwood After a discussion with Pierre, we decided that the code in the driver to assign stream tags for link DMA statically is not suitable to be maintained and therefore it should be removed. This would mean that a new kernel version will be incompatible with an older firmware in terms of HDA functionality.
But we can justify this with the argument that this is broken nevertheless because of the recent programming sequence changes that were introduced.

**Comment by ranj063** [comment]
> @ranj063 approved but you may be ABI MINOR 8, depends on @bardliao PR for trace timestamp.

Thanks @lgirdwood . Fixed ABI minor to 8

File: src/audio/dai.c [hunk 2]
```diff
@@ -423,6 +423,23 @@ static int dai_params(struct comp_dev *dev)
 		return -EINVAL;
 	}
 
+	/* set dma channel */
+	if (dd->chan == DMA_CHAN_INVALID)
+		/* get dma channel at first config only */
+		dd->chan = dma_channel_get(dd->dma, dev->params.stream_tag);
```
Comment (RanderWang) [inline src/audio/dai.c]: Actually dev->params.stream_tag is for host dma, not link dma. And host dma channel is not always equal to link dma channel.

Comment (keyonjie) [reply src/audio/dai.c]: No, this doesn't work, we should add a flag into params for link_stream_tag.

Comment (ranj063) [reply src/audio/dai.c]: @RanderWang @keyonjie for us link DMA channel is always equal to host DMA channel, isnt it? 
Thats the whole argument about why we're having all kinds of issues now. Care to explain your thoughts about a separate link_stream_tag? When would it be different from the host DMA tag? And if it is, how do you make sure that things dont break when you couple them foer setting the DMA format?

Comment (ranj063) [reply src/audio/dai.c]: And before you call out "this doesnt work", I'd appreciate if you could try it out and show me the "not working" usecase. Thanks!

**Comment by ranj063** [comment]
@keqiaozhang can I please request a round of validation tests (including some concurrent usecases) with this PR and corresponding linux PR thesofproject/linux#895 especially on the CML platform?

File: src/audio/dai.c (hunk 2 above)
Comment (keyonjie) [reply src/audio/dai.c]: > And before you call out "this doesnt work", I'd appreciate if you could try it out and show me the "not working" usecase. Thanks!

Sorry, didn't read the whole commit message, now I have more time to write a longer comment, I mean we need solution for M:N in a one shot, and it's good to leave link DMA channel indices allocation/decision from driver side(here your solution is assuming fixing them to host indices and no any flexibility for future).

Can you consider suggestion like below to solve M:N mapping issues: 

```
++struct sof_hda_stream_tag {
++	uint32_t dai_index;	/**< index of this type dai */
++    uint32_t link_stream_tag;
++};

struct sof_ipc_stream_params {
...
	uint32_t rate;
--	uint16_t stream_tag;
++	uint16_t host_stream_tag;
	uint16_t channels;
...
++    uint16_t hda_stream_cnt;
++    struct sof_hda_stream_tag link_stream_tag[0];
} __attribute__((packed));
```
w.r.t. channel indices allocation in driver side, we allocate them in each PCM hw_params, taking 1:N mapping as example, that is, allocate i for host_stream_tag, and i, j, k, ... for the N link_stream_tag array, the couple(to i) --> decouple for stream format setting here is safe here.

OTOH, for M:1 mapping(we have corresponding M FE PCMs in driver side), the 1st (hw_params() running) of M PCM will be allocated to i for host_stream_tag and link_stream_tag, the hw_params() for remaining PCMs only allocate j, k, ... for host_stream_tag in driver side, and in FW side, the HDA dai will only accept the 1st hw_params and ignore the remaining ones. So here it is also safe.

Better to do one more thing in driver side that I talked to Rander, make sure we will couple the channel only at the case that both host channel and link channel are idle, this can improve our robustness for concurrent PCMs, and that is actually recommended from hardware docs.

Comment (ranj063) [reply src/audio/dai.c]: @keyonjie appreciate your thoughts but M:N is quite a complicated case and it cannot be fixed overnight. We do not have the ability in the FW to handle multiple BE's at this point. It needs a larger discussion. I cannot implement a half-baked solution with the hopes that someday it might be useful to fix something. With the current status of HDA platforms, we should resign to the fact that M:N is not supported and will be added later.

Comment (RanderWang) [reply src/audio/dai.c]: @ranj063  for two FEs share one BE,   the dai_param would be called two times and one of them the tag should be different.

Comment (RanderWang) [reply src/audio/dai.c]: And also for some stream with only BE like tone generator,  there is no host stream, right ?

Comment (ranj063) [reply src/audio/dai.c]: @RanderWang are you sure about this? Why would the tag be different? The link DMA ch is already in use and should not be changed when another FE is activated. In this case, the mixer would automatically come into play and it would mix the 2 streams. Why would you change the BE DAI channel?

Comment (ranj063) [reply src/audio/dai.c]: > And also for some stream with only BE like tone generator, there is no host stream, right ?

Agreed, but you would still need to hw_params in this case when you can pass in the stream_tag.

Comment (keyonjie) [reply src/audio/dai.c]: > @keyonjie appreciate your thoughts but M:N is quite a complicated case and it cannot be fixed overnight. We do not have the ability in the FW to handle multiple BE's at this point. It needs a larger discussion. I cannot implement a half-baked solution with the hopes that someday it might be useful to fix something. With the current status of HDA platforms, we should resign to the fact that M:N is not supported and will be added later.

Hi @ranj063 sorry but have to request more than only 1:1 mapping works, at least we should provide the extendable possibility for future, otherwise, we have to go back and forth about it, that consume our effort but no actually improvement from user experience point of view.

**Comment by keqiaozhang** [comment]
> @keqiaozhang can I please request a round of validation tests (including some concurrent usecases) with this PR and corresponding linux PR [thesofproject/linux#895](https://github.com/thesofproject/linux/pull/895) especially on the CML platform?

I verified above PRs on WHL-I2S and GLK, no obvious regressions, but I did run into the IPC times out issue once when doing the volume change stress test( unable to reproduce it again). Besides, these PRs can fix issue https://github.com/thesofproject/linux/issues/854.  

@stevyan, please help to check these PRs on other platforms.

File: src/audio/dai.c (hunk 2 above)
Comment (lgirdwood) [reply src/audio/dai.c]: @keyonjie @RanderWang @ranj063 please make sure the programming flow and ABI is not restricted to 1:1 mappings as other HDA DMA HW may support 1:N or N:1.

Comment (ranj063) [reply src/audio/dai.c]: @lgirdwood i now have some insight into how to make the PR a bit more generic for the M:N case. I will have the updates by tomorrow.

**Comment by Jiangxinx** [comment]
@ranj063 Hi ranjani, I can't add your patch manually, could you help to check it?
```
$ git am 1354.patch
Applying: dai: assign link DMA channel at run-time
error: patch failed: src/include/uapi/abi.h:52
error: src/include/uapi/abi.h: patch does not apply
Patch failed at 0001 dai: assign link DMA channel at run-time
The copy of the patch that failed is found in: .git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```

**Comment by ranj063** [comment]
@plbossart @keyonjie @RanderWang updaed the PR now. I've updated the description with the new method. 

The second commit is quite large but I couldnt separate any parts of it without breaking compilation. Appreciate your comments!

**Comment by ranj063** [comment]
@lgirdwood updated PR now. Instead of setting the link DMA params during dai_params, the driver now sends multiple DAI_CONFIG ipc's to set the link DMA channel. This would scale better for the N:M usecase.

**Comment by keqiaozhang** [comment]
I just run a sanity test on WHL-I2S with this PR, no obvious regressions. I also run some stress tests for it. but I don't have enough time to finish all the tests today, will continue to test it next Monday.

test recipe:
kernel: sof-dev/8663570d2 + PR https://github.com/thesofproject/linux/pull/895
firmware: master/030c81c + PR  #1354

**Comment by kv2019i** [comment]
I tried this on CFL in HDA mode with the scenario I'm debugging for thesofproject/linux#898 and behaviour definitely changes with this patch. Instead of one stream getting stopped (to xrun), I now get a DSP panic with following oops:

linux (91af2f40a699 - ranjani/fix/link-dma-allocation)
fw (f5309a0f - ranjani/fix/link-dma-allocation)

Oops (note invalid channel count at #11):
<pre>
#9  0xbe04c13a in io_reg_read (reg=417633656) at /home/sof/work/sof.git/src/include/sof/io.h:51
#10 host_dma_reg_read (reg=0, chan=4261724474, dma=0x180) at /home/sof/work/sof.git/src/drivers/intel/cavs/hda-dma.c:182
#11 hda_dma_start (dma=0x180, channel=-33242822) at /home/sof/work/sof.git/src/drivers/intel/cavs/hda-dma.c:538
#12 0xbe049750 in dma_start (channel=<optimized out>, dma=<optimized out>) at /home/sof/work/sof.git/src/include/sof/dma.h:263
#13 dai_copy (dev=0xbe0b1500) at /home/sof/work/sof.git/src/audio/dai.c:605
#14 0xbe04250e in comp_copy (dev=0xbe0b1500) at /home/sof/work/sof.git/src/include/sof/audio/component.h:476
#15 pipeline_comp_copy (current=0xbe0b1500, data=<optimized out>, dir=1) at /home/sof/work/sof.git/src/audio/pipeline.c:689
#16 0xbe042bf1 in pipeline_copy (p=0xbe0b1600) at /home/sof/work/sof.git/src/audio/pipeline.c:734
#17 pipeline_task (arg=0xbe0b1600) at /home/sof/work/sof.git/src/audio/pipeline.c:917
</pre>

File: src/audio/dai.c [hunk 3]
```diff
@@ -740,21 +774,23 @@ static int dai_config(struct comp_dev *dev, struct sof_ipc_dai_config *config)
...
+			/* get dma channel at first config only */
+			dd->chan = dma_channel_get(dd->dma, channel);
 
-	/* set up callback */
-	dma_set_cb(dd->dma, dd->chan, DMA_CB_TYPE_IRQ | DMA_CB_TYPE_COPY,
-		   dai_dma_cb, dev);
+		if (dd->chan < 0) {
+			trace_dai_error_with_ids(dev, "dai_config() error: "
+						 "dma_channel_get() failed");
+			return -EIO;
```
Comment (mmaka1) [inline src/audio/dai.c]: Another issue with inconsistent dd->chan value in case of failure.

File: src/audio/dai.c [hunk 4]
```diff
@@ -723,6 +723,40 @@ static int dai_config(struct comp_dev *dev, struct sof_ipc_dai_config *config)
...
+		if (channel != DMA_CHAN_INVALID) {
+			if (dd->chan != DMA_CHAN_INVALID)
+				dma_channel_put(dd->dma, dd->chan);
+
+			dd->chan = dma_channel_get(dd->dma, channel);
+			if (dd->chan < 0) {
+				trace_dai_error_with_ids(dev,
+							 "dai_config() error: "
+							 "dma_channel_get() failed");
+				return -EIO;
```
[the remark this answers is not in the document]
Comment (ranj063) [reply src/audio/dai.c]: Thanks @mmaka1 . makes sense

**Comment by ranj063** [comment]
@mmaka1 your comments addressed now

**Comment by ranj063** [comment]
@kv2019i Thanks for testing. I saw that CI reports the device test on CFL as a failure too. Will investigate it today

**Comment by kv2019i** [comment]
@ranj063 wrote:
> @kv2019i Thanks for testing. I saw that CI reports the device test on CFL as a failure too. Will investigate it today

Actually I deleted the original comment as I was not 100% sure about the results I got and wanted to double-check, but yeah, there were issues.

But good news, with latest linux+sof combo from your fix/link-dma-allocation branch, code now works on CFL, but it also survives the scenario of bug #1413 (previously thesofproject/linux#898 ). I'll continue stresstesting but seems golden.

**Comment by Jiangxinx** [comment]
@ranj063 I have tested this patch with thesofproject/linux#895 on apl-up2-pcm512x platform.
***Test result:***


base | Test result
-- | -- 
Verify firmware presence | Passed
Verify kernel model load probe | Passed
Verify SOF firmware load | Passed
Verify topology binary | Passed
Verify PCM device list | Passed
Check SOF logger | Passed


Analogue Playback Test(hw:0,0) | Test result
-- | -- 
aplay - playback over headset | Passed
stop/resume playback over headset | Passed
Playing -> pause -> s3 -> wakeup -> resume over headset | Passed
repeat 3 times playback over headset | Passed
Pause/Resume playback over headset - 10 times | Passed
Play an audio for 3 minutes | Passed
Suspend & Resume device during audio playback | Passed
Change volume during playback. | Passed


HMDI/DP Playback Test(hw:0, 2) | Test result
-- | -- 
aplay - playback over HDMI | Passed
stop/resume playback over HDMI | Passed
Playing -> pause -> s3 -> wakeup -> resume over HDMI | Passed
repeat 3 times playback over HDMI | Passed
Pause/Resume playback over HDMI | Passed
Play an audio for 3 minutes | Passed
Suspend & Resume device during audio playback | Passed
Change volume during playback. | Passed


Multi-pipeline | Test result
-- | -- 
2 aplay_Analogue + Media Playback | **Failed**
2 aplay_Analogue + HDMI | Passed


General | Test result
-- | -- 
Check the audio channel via speaker-test | Passed
Suspend & Resume device without audio playback or capture | Passed
Check the runtime PM status | Passed
paplay - Play an audio with high-level audio player | Passed
Modules reload | Passed

**Comment by lgirdwood** [comment]
@mmaka1 @keyonjie do you now approve for merge ?
@zrombel CI has been pending a while....

File: src/include/uapi/ipc/dai-intel.h [hunk 5]
```diff
@@ -112,6 +112,7 @@ struct sof_ipc_dai_ssp_params {
 struct sof_ipc_dai_hda_params {
 	struct sof_ipc_hdr hdr;
 	uint32_t link_dma_ch;
+	uint32_t reserved[4];
```
Comment (lyakh) [inline src/include/uapi/ipc/dai-intel.h]: Not sure I understand - how is this related to the fix? Is this the reason to change the ABI minor number? If the answer to both questions is "yes," I think it's better to remove them from this patch

Comment (ranj063) [reply src/include/uapi/ipc/dai-intel.h]: @lyakh this is a pretend ABI update to track this change. With this update to link DMA assignment, the kernel/FW are officially backwards incompatible

Comment (lyakh) [reply src/include/uapi/ipc/dai-intel.h]: For incompatibility we have to increment the major version, don't we? Still you don't need to add those reserved fields, if they're unrelated.

Comment (ranj063) [reply src/include/uapi/ipc/dai-intel.h]: @lyakh that was my original change but this is what has been agreed with maintainers. A major ABI change usually requires big changes to the IPC structs. I was told functional incompatibility cannot be classified a major ABI change. So, we advertise full HDA support only 3.8 onwards.

**Comment by lgirdwood** [comment]
@mmaka1 @keyonjie @lyakh any blockers or can this be merged ?

**Comment by ranj063** [comment]
@lgirdwood could we please hold off merging this one until stress tests are complete and both the kernel/fw patches can go in together or else all CI tests will fail.

**Comment by lgirdwood** [comment]
@ranj063 ok, just ping me when we are good.

**Comment by jajanusz** [comment]
@mmaka1 @keyonjie please re-review

**Comment by lgirdwood** [comment]
@plbossart we are good to go here, with the exception of the ABI major bump. Since we are NOT changing ABI, and only changing HDA programming flow, we could use version number (i.e. v1.3 and above) to use new HDA programming flow and v1.2 and below for old flow ?

**Comment by lgirdwood** [comment]
@ranj063 could you revert the ABI update part if we agree on version number to change HDA programming flow.

**Comment by xiulipan** [comment]
Manually run with latest https://github.com/thesofproject/linux/pull/895 binary

**Comment by plbossart** [comment]
> @ranj063 could you revert the ABI update part if we agree on version number to change HDA programming flow.

I don't want to even maintain the old flow. It's just broken so we only need to claim HDaudio is supported starting with release 1.3

File: src/include/uapi/ipc/dai-intel.h (hunk 5 above)
Comment (plbossart) [reply src/include/uapi/ipc/dai-intel.h]: @ranj063 this and the ABI bump are not needed.

**Comment by ranj063** [comment]
> @ranj063 could you revert the ABI update part if we agree on version number to change HDA programming flow.

@plbossart @lgirdwood sure. I will update both the kernel/FW PR's to remove the ABI bump by EOD

**Comment by ranj063** [comment]
@lgirdwood @plbossart updated to remove ABI change

File: src/audio/dai.c (hunk 3 above)
Comment (plbossart) [reply src/audio/dai.c]: @ranj063 did you address Marcin's comment?

Comment (ranj063) [reply src/audio/dai.c]: @plbossart yes. he has already approved the PR as well.

**Comment by plbossart** [comment]
matching PR merged on kernel side.

**Comment by jajanusz** [comment]
@zrombel Please check CI

**Comment by lgirdwood** [comment]
@jajanusz thanks, we can merge now as soon as CI is done.
